main/lsp/LSPTypecheckerCoordinator.cc

```diff
@@ -32,59 +106,28 @@ void LSPTypecheckerCoordinator::asyncRunInternal(shared_ptr<core::lsp::Task> tas
     }
 }
 
-void LSPTypecheckerCoordinator::syncRun(function<void(LSPTypecheckerDelegate &)> &&lambda, bool multithreaded) {
+void LSPTypecheckerCoordinator::syncRun(unique_ptr<LSPTask> task, bool multithreaded) {
     absl::Notification notification;
-    CounterState typecheckerCounters;
-    // If typechecker is running on a dedicated thread, then we need to merge its metrics w/ coordinator thread's so we
-    // report them.
-    // Note: Capturing notification by reference is safe here, we we wait for the notification to happen prior to
-    // returning.
-    asyncRunInternal(make_shared<LambdaTask>(
-        [&typechecker = this->typechecker, &workers = multithreaded ? workers : *emptyWorkers, lambda, &notification,
-         &typecheckerCounters, hasDedicatedThread = this->hasDedicatedThread]() -> void {
-            LSPTypecheckerDelegate d(workers, typechecker);
-            lambda(d);
-            if (hasDedicatedThread) {
-                typecheckerCounters = getAndClearThreadCounters();
-            }
-            notification.Notify();
-        }));
-    notification.WaitForNotification();
-    if (hasDedicatedThread) {
-        counterConsume(move(typecheckerCounters));
-    }
+    auto wrappedTask = make_shared<TypecheckerTask>(
+        move(task), make_unique<LSPTypecheckerDelegate>(workers, typechecker), hasDedicatedThread);
+    asyncRunInternal(wrappedTask);
+    wrappedTask->blockUntilComplete();
 }
 
 void LSPTypecheckerCoordinator::initialize(unique_ptr<InitializedParams> params) {
     // TODO: Make async when we land preemptible slow path.
-    syncRun([&typechecker = this->typechecker, &workers = workers, &params = params](auto &tcd) -> void {
-        auto &updates = params->updates;
-        typechecker.initialize(move(updates), workers);
-    });
+    syncRun(make_unique<InitializeTask>(*config, typechecker, workers, move(params->updates)));
```

Comment:
Discussed this comment offline – we should be careful because part of `InitializeTask` probably needs to remain `sync` (to make sure subsequent tasks aren't given an `LSPTypecheckerDelegate` that isn't ready for use), but yes, part of it can be made `async`. 

 
TBD in a future commit how to implement that, for example, maybe by splitting the current InitializeTask into separate sync and async sub-tasks?

Reply:
Added a comment about this, and some light documentation on the new classes.